upstairs/src/volume.rs

```diff
@@ -610,6 +610,72 @@ impl BlockIO for Volume {
 
         Ok(())
     }
+    async fn activate_with_gen(&self, gen: u64) -> Result<(), CrucibleError> {
+        for sub_volume in &self.sub_volumes {
+            sub_volume.conditional_activate_with_gen(gen).await?;
+
+            let sub_volume_computed_size = self.block_size
+                * (sub_volume.lba_range.end - sub_volume.lba_range.start);
+
+            if sub_volume.total_size().await? != sub_volume_computed_size {
+                crucible_bail!(SubvolumeSizeMismatch);
+            }
+        }
+
+        if let Some(ref read_only_parent) = &self.read_only_parent {
+            read_only_parent.conditional_activate_with_gen(gen).await?;
+        }
+
+        Ok(())
+    }
+
+    // Return a vec of these?
+    // Placeholder for first round of crutest changes.
+    // I think this should return some sort of VolumeInfo struct with an
+    // SV/ROP structure where the SV is a Vec<WQC>  and the ROP is Option<WQC>
+    async fn query_work_queue(&self) -> Result<WQCounts, CrucibleError> {
+        let mut all_wq = WQCounts {
+            up_count: 0,
+            ds_count: 0,
+            active_count: 0,
+        };
+
+        for sub_volume in &self.sub_volumes {
+            let sv_wq = sub_volume.query_work_queue().await?;
+            all_wq.up_count += sv_wq.up_count;
+            all_wq.ds_count += sv_wq.ds_count;
+            all_wq.active_count += sv_wq.active_count;
+        }
+
+        if let Some(ref read_only_parent) = &self.read_only_parent {
+            let rop_wq = read_only_parent.query_work_queue().await?;
+            all_wq.up_count += rop_wq.up_count;
+            all_wq.ds_count += rop_wq.ds_count;
+            all_wq.active_count += rop_wq.active_count;
+        }
+
+        Ok(all_wq)
+    }
+
+    // Return a vec of these?
+    // Return a struct with a vec for SV and Some/None for ROP?
+    async fn query_extent_size(&self) -> Result<Block, CrucibleError> {
```

Comment:
What's this used for? I wonder why the caller is ask and if that could be changed.

Reply:
We use it in crutest to construct specific kinds of tests where we cross the extent boundary with IOs.  It could be done other ways, like we pass the test itself information externally.  Having the test figure it out without having to supply anything is an easy way to do it.

Reply:
The `sparse_fill()` is another good example.  We want to do an IO to every extent so they all end up dirty, but dont' want to have to do IO to every block in the extent.